File: solutions/java/0034.java
/*
 * Leetcode 34: Find First and Last Position of Element in Sorted Array. Java. Medium.
 * O(log n) runtime and O(1) space complexity using two binary searches.
 * The problem is solved by first finding the leftmost occurrence of the target with a binary search
 * and then performing a second binary search to find the rightmost occurrence.
 * Edge cases such as an empty array or target not present are naturally handled.
 * 12/5/2024 Winston Tsui
 */

// TODO: Redo solution with one binary search instead of two.
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0)
            return new int[]{-1, -1};

        int l = 0, r = nums.length - 1;

        // Find first target value
        while (l < r) {
            int m = l + (r - l) / 2;
            if (nums[m] >= target)
                r = m;
            else
                l = m + 1;
        }

        // If target is not found
        if (nums[r] != target)
            return new int[]{-1, -1};
        int start = r;

        // Find last target value
        l = r; // Start from the first occurrence of target
        r = nums.length - 1;
        while (l < r) {
            int m = l + (r - l + 1) / 2; // Adjust midpoint to favor the right
            if (nums[m] > target)
                r = m - 1;
            else
                l = m; // Keep moving to the right
        }

        return new int[]{start, l};
    }
}

File: solutions/python/0034.py
"""
 * Leetcode 34: Find First and Last Position of Element in Sorted Array. Python. Medium.
 * O(log N) runtime, O(1) space complexity
 * Use two binary searches to find first and last occurrence of target.
 *
 * Input: nums = [5,7,7,8,8,10], target = 8
 * Output: [3,4]
 *
 * Input: nums = [5,7,7,8,8,10], target = 6
 * Output: [-1,-1]
 *
 * 6/28/2025 Winston Tsui
"""

from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def find_first():
            l, r = 0, len(nums) - 1
            found = False

            while l <= r:
                m = l + (r - l) // 2

                if nums[m] >= target:
                    r = m - 1
                else:
                    l = m + 1
            return l if l < len(nums) and nums[l] == target else -1

        def find_second():
            l, r = 0, len(nums) - 1

            while l <= r:
                m = l + (r - l) // 2

                if nums[m] > target:
                    r = m - 1
                else:
                    l = m + 1
            return r if r >= 0 and nums[r] == target else -1

        l = find_first()
        r = find_second()

        return [l, r]
